proxy/src/auth/credentials.rs

```diff
@@ -40,49 +43,319 @@ impl ClientCredentials {
     }
 }
 
-#[derive(Debug, Error)]
+#[derive(Debug, Error, PartialEq)]
 pub enum ProjectNameError {
     #[error("SNI is missing. EITHER please upgrade the postgres client library OR pass the project name as a parameter: '...&options=project%3D<project-name>...'.")]
     Missing,
 
-    #[error("SNI is malformed.")]
-    Bad,
+    #[error("Inconsistent project name inferred from SNI and project option. Project name from SNI: '{0}', Project name from project option: '{1}'")]
+    InconsistentProjectNameAndSNI(String, String),
+
+    #[error("Common name is not set.")]
+    CommonNameNotSet,
+
+    #[error("Inconsistent common name and SNI suffix. Common name: '{0}', SNI: '{1}', SNI suffix: '{2}'")]
+    InconsistentCommonNameAndSNI(String, String, String),
 
-    #[error("Inconsistent project name inferred from SNI and project option. String from SNI: '{0}', String from project option: '{1}'")]
-    Inconsistent(String, String),
+    #[error("Project name must contain only alphanumeric characters and hyphens ('-'). Project name: '{0}'.")]
+    ProjectNameContainsIllegalChars(String),
 }
 
 impl UserFacingError for ProjectNameError {}
 
-impl ClientCredentials {
-    /// Determine project name from SNI or from project_name parameter from options argument.
-    pub fn project_name(&self) -> Result<&str, ProjectNameError> {
-        // Checking that if both sni_data and project_name are set, then they should match
-        // otherwise, throws a ProjectNameError::Inconsistent error.
-        if let Some(sni_data) = &self.sni_data {
-            let project_name_from_sni_data =
-                sni_data.split_once('.').ok_or(ProjectNameError::Bad)?.0;
-            if let Some(project_name_from_options) = &self.project_name {
-                if !project_name_from_options.eq(project_name_from_sni_data) {
-                    return Err(ProjectNameError::Inconsistent(
-                        project_name_from_sni_data.to_string(),
-                        project_name_from_options.to_string(),
-                    ));
+/// Inferring project name from sni_data.
+/// If common_name is None, throws an CommonNameNotSet error,
+/// otherwise checks if common_name is suffix to sni_data.
+/// if common_name is not suffix to sni_data throws InconsistentCommonNameAndSNI error,
+/// otherwise returns the prefix of sni_data without the common_name suffix.
+fn project_name_from_sni_data<'sni>(
+    sni_data: &'sni str,
+    common_name: &Option<String>,
+) -> Result<&'sni str, ProjectNameError> {
+    // extract common name. If unset, throw a CommonNameNotSet error.
+    let common_name = match &common_name {
+        Some(common_name) => common_name,
+        None => return Err(ProjectNameError::CommonNameNotSet),
+    };
```

Comment:
Sorry, I'm confused. Why do you take `Option` if not having a value is an immediate error?

Reply:
Good catch. 
common_name is an option from the start (in main), because it's derived from server.crt, which is also an option. So it could happen that common_name is None, and since we have to handle this at some point, I decided to handle it here.